test/bats/test.bats

```diff
@@ -0,0 +1,114 @@
+  waitForProcess "60" "1" "$cmd"
+
+  run kubectl -n gatekeeper-system get crd/constrainttemplates.templates.gatekeeper.sh
+  assert_success
+}
+
+@test "waiting for validating webhook" {
+  cmd="kubectl get validatingwebhookconfigurations.admissionregistration.k8s.io validation.gatekeeper.sh"
+	waitForProcess "60" "1" "$cmd"
+
+  run kubectl get validatingwebhookconfigurations.admissionregistration.k8s.io validation.gatekeeper.sh
+  assert_success
+}
+
+@test "applying sync config" {
+  run kubectl apply -f ${BATS_TESTS_DIR}/sync.yaml
+  assert_success
+}
+
+# creating a namespace early so it would sync
+@test "create namespace for unique labels test" {
+  run kubectl apply -f ${BATS_TESTS_DIR}/good/no_dupe_ns.yaml
+  assert_success
+}
+
+@test "container limits test" {
+  run kubectl apply -f ${BATS_TESTS_DIR}/templates/k8scontainterlimits_template.yaml
+  assert_success
+
+  cmd="kubectl -n gatekeeper-system wait --for condition=established --timeout=60s crd/k8scontainerlimits.constraints.gatekeeper.sh"
+  waitForProcess "60" "1" "$cmd"
+
+  run kubectl get crd/k8scontainerlimits.constraints.gatekeeper.sh
+  assert_success
+
+  run kubectl apply -f ${BATS_TESTS_DIR}/constraints/containers_must_be_limited.yaml
+  assert_match 'k8scontainerlimits.constraints.gatekeeper.sh/container-must-have-limits created' "$output"
+  assert_success
+
+  cmd="kubectl get k8scontainerlimits.constraints.gatekeeper.sh container-must-have-limits -o yaml | grep enforced"
+  waitForProcess "60" "1" "$cmd"
+
+  run kubectl create ns production
+  assert_success
+
+  run kubectl apply -f ${BATS_TESTS_DIR}/bad/opa_no_limits.yaml
+  assert_match 'denied the request' "$output"
+  assert_failure
+
+  run kubectl apply -f ${BATS_TESTS_DIR}/good/opa.yaml
+  assert_success
+}
+
+@test "required labels test" {
+  run kubectl apply -f ${BATS_TESTS_DIR}/templates/k8srequiredlabels_template.yaml
+  assert_success
+
+  cmd="kubectl -n gatekeeper-system wait --for condition=established --timeout=60s crd/k8srequiredlabels.constraints.gatekeeper.sh"
+	waitForProcess "60" "1" "$cmd"
+
+  run kubectl apply -f ${BATS_TESTS_DIR}/constraints/all_ns_must_have_gatekeeper.yaml
+  assert_success
+
+  cmd="kubectl get k8srequiredlabels.constraints.gatekeeper.sh ns-must-have-gk -o yaml | grep enforced"
+  waitForProcess "60" "1" "$cmd"
+
+  run kubectl apply -f ${BATS_TESTS_DIR}/good/good_ns.yaml
+  assert_success
+
+  run kubectl apply -f ${BATS_TESTS_DIR}/bad/bad_ns.yaml
+  assert_match 'denied the request' "$output"
+  assert_failure
+
+  # waiting for audit interval (default is 60s)
+  sleep 60
+
+  result=$(kubectl get k8srequiredlabels.constraints.gatekeeper.sh ns-must-have-gk -o json | jq '.status.violations[] | length' | head -1)
+  [[ "$result" -eq 4 ]]
+}
+
+@test "unique labels test" {
+  run kubectl apply -f ${BATS_TESTS_DIR}/templates/k8suniquelabel_template.yaml
```

Comment:
We can make sure that the resource has been synced by checking this command:

`kubectl get ns gatekeeper-system -o jsonpath='{.metadata.finalizers}{"\n"}' | grep finalizers.gatekeeper.sh/sync`

Reply:
thanks, added before uniqueness test

Reply:
thanks, added before uniqueness test